extensions/amp-render/1.0/amp-render.js

```diff
@@ -103,86 +136,49 @@ class AmpRender extends BaseElement {
-                  .renderTemplateAsString(dev().assertElement(template), data)
-                  .then((html) => dict({'__html': html}));
-              },
-            })
-          );
-        });
-      } else {
-        this.mutateProps(dict({'render': null}));
-      }
+    if (template === this.template_) {
+      return;
+    }
+    this.template_ = template;
+    if (template) {
+      // Only overwrite `render` when template is ready to minimize FOUC.
+      templates.whenReady(template).then(() => {
+        if (template != this.template_) {
+          // A new template has been set while the old one was initializing.
+          return;
+        }
+        this.mutateProps(
+          dict({
+            'render': (data) => {
+              return templates
+                .renderTemplateAsString(dev().assertElement(template), data)
+                .then((html) => dict({'__html': html}));
+            },
+          })
+        );
+      });
+    } else {
+      this.mutateProps(dict({'render': null}));
     }
   }
 
   /**
-   * TODO: this implementation is identical to one in amp-data-display &
+   * TODO: this implementation is identical to one in amp-date-display &
    * amp-date-countdown. Move it to a common file and import it.
    *
    * @override
    */
   isReady(props) {
-    if (this.template_ && !('render' in props)) {
-      // The template is specified, but not available yet.
-      return false;
-    }
-    return true;
-  }
-
-  /**
-   * Gets the json an amp-list that has an "amp-state:" uri. For example,
-   * src="amp-state:json.path".
-   *
-   * TODO: this implementation is identical to one in amp-list. Move it
-   * to a common file and import it
-   *
-   * TODO: Add src as a param once this method is moved out of this class
-   * to support src binding (https://github.com/ampproject/amphtml/pull/33189#discussion_r598743971).
-   *
-   * @param {!AmpElement} element
-   * @return {Promise<!JsonObject>}
-   */
-  getAmpStateJson(element) {
-    const src = element.getAttribute('src');
-    if (!src) {
-      return Promise.resolve({});
-    }
-    return Services.bindForDocOrNull(element)
-      .then((bind) => {
-        userAssert(bind, '"amp-state:" URLs require amp-bind to be installed.');
-        const ampStatePath = src.slice(AMP_STATE_URI_SCHEME.length);
-        return bind.getStateAsync(ampStatePath).catch((err) => {
-          const stateKey = ampStatePath.split('.')[0];
-          user().error(
-            TAG,
-            `'amp-state' element with id '${stateKey}' was not found.`
-          );
-          throw err;
-        });
-      })
-      .then((json) => {
-        userAssert(
-          typeof json !== 'undefined',
-          `[amp-render] No data was found at provided uri: ${src}`
-        );
-        return json;
-      });
+    // If a template is specified, then it must be available.
+    return !this.template_ || 'render' in props;
   }
 }
 
+AmpRender['props'] = {
```

Comment:
FMI: does props collection happen each time an attribute on the `<amp-render>` is modified? I.e. would a user script modifying the element's `src` result in `getJsonFn` being called again?

Reply:
That's a good question. Yes, `getJsonFn` does get called every time `src` is modified. However, since it's not doing any heavy computation, doesn't make sense to cache the result.